### PR TITLE
Adds reinterpret_const with a test.

### PR DESCRIPTION
**Pull request by jiawen**
Closes: #85.

### CODE REVIEW

File: include/array/array.h [hunk 1]
```diff
@@ -1252,7 +1252,8 @@ NDARRAY_HOST_DEVICE auto transpose_impl(const Shape& shape, index_sequence<Extra
  *
  * Examples:
  * - `transpose<2, 0, 1>(s) == make_shape(s.dim<2>(), s.dim<0>(), s.dim<1>())`
- * - `transpose<1, 0>(s) == make_shape(s.dim<1>(), s.dim<0>(), ...)` where ... is all dimensions after dimension 1. */
+ * - `transpose<1, 0>(s) == make_shape(s.dim<1>(), s.dim<0>(), ...)` where ... is all dimensions
```
Comment (jiawen) [inline include/array/array.h:1255]: My editor ran `clang-format` on its own. LMK if you want to revert.

File: test/BUILD.bazel [hunk 2]
```diff
@@ -4,6 +4,7 @@ cc_test(
     srcs = [
         "algorithm.cpp",
         "array.cpp",
+        "array_ref.cpp",
```
Comment (jiawen) [inline test/BUILD.bazel:7]: Oops.

File: test/array_ref.cpp [hunk 3]
```diff
@@ -87,8 +104,8 @@ TEST(array_ref_incompatible_shape) {
   }
 
   {
-    array_ref<int, nda::shape<nda::fixed_dim<10, nda::dynamic>,
-                              nda::fixed_dim<2, nda::dynamic>>> dst;
+    array_ref<int, nda::shape<nda::fixed_dim<10, nda::dynamic>, nda::fixed_dim<2, nda::dynamic>>>
```
Comment (jiawen) [inline test/array_ref.cpp:107]: Ugly - want me to revert this?

Comment (dsharlet) [reply test/array_ref.cpp:107]: I agree but I wouldn't bother if it's what clang-format will do every time.

File: include/array/array.h [hunk 4]
```diff
@@ -2890,6 +2891,13 @@ const_array_ref<T, NewShape> reinterpret_shape(
   return reinterpret_shape(a.cref(), new_shape, offset);
 }
 
+/** Reinterpret a const_array_ref as a array_ref. This "reinterprets away the
+ the const-ness on T" and is analogous to const_cast. */
+template <class T, class Shape>
+array_ref<T, Shape> reinterpret_const(const_array_ref<T, Shape>& a) {
```
[the remark this answers is not in the document]
Comment (jiawen) [reply include/array/array.h:2897]: Oh, I see. I can try that.

**Comment by jiawen** [comment]
@dsharlet How about this? It's a little "too easy to use" but it's also named `reinterpret` so it seems fine to me.

All the tests pass.

Any remaining gotchas?

**Comment by dsharlet** [comment]
I think the readability of this version is a bit questionable, if you read something like `a = reinterpret(b)`, it's a bit surprising that this is removing `const`. 

I think if we're going to keep this API, it should be named `reinterpret_const` despite the fact that it's a new function instead of an overload (which I do like).

I still think the API that is the same as the existing `reinterpret` except it uses `const_cast` instead of `reinterpret_cast` internally makes the most sense. That way, `reinterpret`, `reinterpret_cast`, `const_cast`, and `reinterpret_const` all have the same basic API and usage (explicitly specified new type).

FWIW I have always thought it was a bit of a wart that `const_cast` needed the new type (not just constness) even though its redundant, but I think we should mirror that wart in our API.

File: include/array/array.h [hunk 5]
```diff
@@ -2870,6 +2871,13 @@ const_array_ref<U, Shape> reinterpret(const array<T, Shape, Alloc>& a) {
   return reinterpret<const U>(a.cref());
 }
 
+/** Reinterpret a const_array_ref as a array_ref. This "reinterprets away the
```
Comment (dsharlet) [inline include/array/array.h:2874]: I think the comment is a bit stale from older versions maybe?

File: test/errors.cpp [hunk 6]
```diff
@@ -148,6 +146,15 @@ void make_move_different_rank() { auto a2 = make_move(a, shape_of_rank<2>()); }
 
 void make_move_ref_different_rank() { auto a2 = make_move(ref, shape_of_rank<3>()); }
 
+// When commented out, this fails to compile as expected, but since the error
```
Comment (dsharlet) [inline test/errors.cpp:149]: Uncommented?

We *could* add an appropriate `enable_if` to `reinterpret_const` to make this test pass. I might try to do that at some point, it’s not necessary for now.

**Review by dsharlet — APPROVED** [review]

File: include/array/array.h (hunk 5 above)
Comment (jiawen) [reply include/array/array.h:2874]: How about:

```
/** Reinterpret the const_array_ref `a` of type `T` (aka array_ref<const T>) to have a different type
 * `U` using `const_cast`.
 */
```

File: test/errors.cpp (hunk 6 above)
Comment (jiawen) [reply test/errors.cpp:149]: Will do when I get home to a computer that has my ssh keys!

File: include/array/array.h (hunk 5 above)
Comment (dsharlet) [reply include/array/array.h:2874]: LGTM

File: test/errors.cpp (hunk 6 above)
Comment (jiawen) [reply test/errors.cpp]: Done.

**Pull request merged**